std/machines/small_field/rotate.asm

```diff
@@ -0,0 +1,76 @@
+use std::utils::unchanged_until;
+use std::utils::cross_product;
+use std::convert::int;
+use std::check::require_field_bits;
+
+/// Rotate on single byte input using an exhaustive table, returning two 16-bit values as result.
+/// We can rotate by at most 31 bits
+machine ByteRotate with
+    latch: latch,
+    operation_id: operation_id,
+    degree: 65536
+{
+    // P_C0 and P_C1 are both 16 bit limbs of P_C, where P_C0 is the less significant limb.
+    operation run<0> P_operation, P_A, P_B, P_ROW -> P_C0, P_C1;
+    col fixed latch = [1]*;
+    col fixed operation_id = [0]*;
+
+    require_field_bits(16, || "The field modulo should be at least 2^16 - 1 to work in the rotate machine.");
+
+    let bit_counts = [256, 32, 4, 2];
+    let min_degree = std::array::product(bit_counts);
+    std::check::assert(std::prover::min_degree() >= std::array::product(bit_counts), || "The rotate machine needs at least 65536 rows to work.");
+    
+    let out_byte_offset = cross_product(bit_counts);
+    let a: int -> int = out_byte_offset[0];
+    let b: int -> int = out_byte_offset[1];
+    let row: int -> int = out_byte_offset[2];
+    let op: int -> int = out_byte_offset[3];
+
+    let P_A: col = a;
+    let P_B: col = b;
+    let P_ROW: col = row;
+    let P_operation: col = op;
+    let P_C: int -> int = |i| match op(i) {
+        0 => a(i) << (b(i) + (row(i) * 8)) | (a(i) << (row(i) * 8)) >> (32 - b(i)),
+        1 => ((a(i) << (row(i) * 8)) >> b(i)) | (a(i) << (32 - b(i) + (row(i) * 8))),
+    };
+    col fixed P_C0(i) { P_C(i) & 0xffff };
+    col fixed P_C1(i) { (P_C(i) >> 16) & 0xffff };
+}
+
+machine Rotate(byte_rotate: ByteRotate) with
+    latch: latch,
+    operation_id: operation_id,
+    // Allow this machine to be connected via a permutation
+    call_selectors: sel,
+{
+    require_field_bits(17, || "Rotate requires a field that fits any 17-Bit value.");
```

Comment:
Why does this need 17 and the sub-machine only 16?

Reply:
Hmm, good question, I'm not sure actually. I supposed it's a requirement as it is in the shift machine

Reply:
@leonardoalt I think you added the 17 for the shift machine - are you sure about it?

Reply:
https://github.com/powdr-labs/powdr/blob/main/std/machines/small_field/shift.asm#L48

Reply:
I thought it is related to f17 f48 thing which I don't know about it

Reply:
Arith also has 48 bits requirement

Reply:
https://github.com/powdr-labs/powdr/blob/main/std/machines/large_field/arith.asm#L25